addons/binding/org.openhab.binding.rotelra1x/src/main/java/org/openhab/binding/rotelra1x/handler/RotelRa1xHandler.java

```diff
@@ -0,0 +1,350 @@
+        if (serialPort != null) {
+            serialPort.close();
+        }
+        serialPort = null;
+    }
+
+    private String readCommand() throws IOException {
+        return readUntil('!', '=');
+    }
+
+    private String readUntil(char terminator) throws IOException {
+        return readUntil(terminator, terminator);
+    }
+
+    private String readUntil(char terminator1, char terminator2) throws IOException {
+        StringBuilder result = new StringBuilder();
+        while (true) {
+            int b = serialPort.getInputStream().read();
+            if (b == -1) {
+                throw new IOException("Connection unexpectedly closed");
+            } else if (b == terminator1 || b == terminator2) {
+                break;
+            } else {
+                result.append((char) b);
+            }
+        }
+        return result.toString();
+    }
+
+    public PercentType readVolume() throws IOException {
+        String volumeString = readUntil('!');
+        int volume;
+        if ("min".equals(volumeString)) {
+            volume = 0;
+        } else if ("max".equals(volumeString)) {
+            volume = maximumVolume;
+        } else {
+            volume = Integer.parseInt(volumeString, 10);
+        }
+        double volumePct = volume * 100.0 / maximumVolume;
+        return new PercentType(BigDecimal.valueOf(Math.round(volumePct)));
+    }
+
+    public PercentType readDimmer() throws IOException {
+        String dimmerString = readUntil('!');
+        int dimmer = Integer.parseInt(dimmerString, 10);
+        // Let's make 100 % the brightest, makes more sense
+        double dimmerPct = 100.0 - (dimmer * 100.0 / 6.0);
+        return new PercentType(BigDecimal.valueOf(Math.round(dimmerPct)));
+    }
+
+    public DecimalType readFrequency() throws IOException {
+        String freqString = readUntil('!');
+        double freq;
+        if ("off".equals(freqString)) {
+            freq = 0.0;
+        } else {
+            freq = Double.parseDouble(freqString);
+        }
+        return new DecimalType(freq);
+    }
+
+    void powerOnRefresh() {
+        scheduler.schedule(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    send("get_volume!");
+                    send("get_current_source!");
+                } catch (IOException | ConfigurationError e) {
+                    logger.info("Failed to request volume and source after powering on.", e);
+                }
+            }
+        }, 0, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void run() {
+        while (!exit) {
+            try {
+                if (serialPort == null) {
+                    connect();
+                }
+                String command = readCommand();
+                if ("volume".equals(command)) {
+                    PercentType vol = readVolume();
+                    updateState(getThing().getChannel("volume").getUID(), vol);
```

Comment:
You should not use Strings like "volume" here - better define constants for your channel ids.
And then you can take the shortcut:
```
```updateState(CHANNEL_VOLUME, vol);